code/android-core-library/src/phone/java/com/adobe/marketing/mobile/services/caching/CacheFileManager.java

```diff
@@ -0,0 +1,290 @@
+import com.adobe.marketing.mobile.internal.util.FileUtils;
+import com.adobe.marketing.mobile.internal.util.StringEncoder;
+import com.adobe.marketing.mobile.services.DeviceInforming;
+import com.adobe.marketing.mobile.services.Log;
+import com.adobe.marketing.mobile.services.ServiceConstants;
+import com.adobe.marketing.mobile.util.StringUtils;
+
+import org.json.JSONException;
+import org.json.JSONObject;
+import org.json.JSONTokener;
+
+import java.io.ByteArrayInputStream;
+import java.io.File;
+import java.io.InputStream;
+import java.nio.charset.StandardCharsets;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+
+/**
+ * Helper class for {@code FileCacheService} used for managing cache files and their metadata.
+ */
+class CacheFileManager {
+    private static final String TAG = "CacheFileManager";
+
+    @VisibleForTesting
+    static final String CACHE_METADATA_FILE_EXT = "_metadata.txt";
+
+    private final DeviceInforming deviceInfoService;
+    private final String rootCacheDirName;
+
+    CacheFileManager(@NonNull final DeviceInforming deviceInfoService,
+                     @NonNull final String rootCacheDirName) {
+        this.rootCacheDirName = rootCacheDirName;
+        this.deviceInfoService = deviceInfoService;
+    }
+
+    /**
+     * Creates a directory with name {@code cacheName} inside the {@code cacheRoot}
+     * folder of {@code deviceInfoService.getApplicationCacheDir()}
+     *
+     * @param cacheName the name of the folder to act as a cache bucket.
+     * @return a {@code File} handle for the newly created/existing bucket;
+     * null if the directory creation fails.
+     */
+    File createCache(final String cacheName) {
+        if (StringUtils.isNullOrEmpty(cacheName)) return null;
+
+        final File appCacheDir = deviceInfoService.getApplicationCacheDir();
+        if (!FileUtils.isWritableDirectory(appCacheDir)) {
+            Log.debug(ServiceConstants.LOG_TAG, TAG, "App cache directory is not writable.");
+            return null;
+        }
+
+        final File cacheBucket = new File(appCacheDir, rootCacheDirName + File.separator + cacheName);
+        if (!cacheBucket.exists() && !cacheBucket.mkdirs()) {
+            Log.debug(ServiceConstants.LOG_TAG, TAG, "Cannot create cache bucket.");
+            return null;
+        }
+
+        return cacheBucket;
+    }
+
+    /**
+     * Retrieves a cache file for {@code key} from {@code cacheName} directory.
+     *
+     * @param cacheName the cache sub-folder within {@code cacheRoot} where
+     *                  cache file stored for key should be fetched from
+     * @param key       the key for which the cache file should be retrieved
+     * @return the cache file for {@code key} if present; null otherwise.
+     */
+    File getCacheFile(final String cacheName, final String key) {
+        final String entryLocation = getCacheLocation(cacheName, key);
```

Comment:
Missing null checks

Reply:
The null checks are delegated to `getCacheLocation()`. Will make them `@NonNull` and do them in the public entry api.